src/Npgsql/TypeMapping/ConnectorTypeMapper.cs

```diff
@@ -380,20 +380,12 @@ void BindRangeType(NpgsqlTypeHandler elementHandler, PostgresRangeType pgRangeTy
                 ? NpgsqlDbType.Range | elementNpgsqlDbType.Value
                 : (NpgsqlDbType?)null;
 
+            // We only want to bind supported range CLR types whose element CLR types are being bound as well.
+            var clrTypes = rangeHandler.SupportedElementClrTypes
+                .Where(t => elementClrTypes.Contains(t.GenericTypeArguments[0]))
+                .ToList();
 
-            Type[]? clrTypes = null;
-            if (elementClrTypes != null)
-            {
-                // Somewhat hacky. Although the element may have more than one CLR mapping,
-                // its range will only be mapped to the "main" one for now.
-                var defaultElementType = elementHandler.GetFieldType();
-
-                clrTypes = elementClrTypes.Contains(defaultElementType)
-                    ? new[] { rangeHandler.GetFieldType() }
-                    : null;
-            }
-
-            BindType(rangeHandler, pgRangeType, rangeNpgsqlDbType, null, clrTypes);
+            BindType((NpgsqlTypeHandler)rangeHandler, pgRangeType, rangeNpgsqlDbType, null, clrTypes.ToArray());
```

Comment:
> Regarding perf, this is one of those places where I think we should prefer design correctness and codebase quality over such a micro-optimization.

But this doesn't look as quality code because of the cast.

Reply:
What exactly are you proposing? That we go back to the current situation (removing 0faeb5e5d1348e819a6ff0321000bd7928700c73), making RangeHandler<T> extend an abstract non-generic RangeHandler as before, and duplicating the methods from NpgsqlTypeHandler<T> (e.g. WriteObjectWithLength)? That doesn't make sense to me...

Reply:
I can make the cast a bit nicer with a check and throw, if that's what's bothering you...